File: src/constants/badges.ts
import type { Badge } from '../types';

export const BADGES: Badge[] = [
  { badgeId: '0', level: 0, name: 'Observador Anônimo', nameEN: 'Anonymous Observer', icon: '👁️', minReputation: 0, maxReputation: 49, dailyReportLimit: 5, color: '#6B7280', glowColor: '#374151', perks: ['basic_map', 'create_reports'] },
  { badgeId: '1', level: 1, name: 'Cidadão Alerta', nameEN: 'Alert Citizen', icon: '🔔', minReputation: 50, maxReputation: 99, dailyReportLimit: 5, color: '#9CA3AF', glowColor: '#4B5563', perks: ['confirm_deny'] },
  { badgeId: '2', level: 2, name: 'Vigia Iniciante', nameEN: 'Beginner Watchman', icon: '🔍', minReputation: 100, maxReputation: 199, dailyReportLimit: 5, color: '#60A5FA', glowColor: '#2563EB', perks: ['comments'] },
  { badgeId: '3', level: 3, name: 'Repórter de Rua', nameEN: 'Street Reporter', icon: '📝', minReputation: 200, maxReputation: 349, dailyReportLimit: 8, color: '#34D399', glowColor: '#059669', perks: ['photos', '8_reports'] },
  { badgeId: '4', level: 4, name: 'Olheiro Comunitário', nameEN: 'Community Scout', icon: '👀', minReputation: 350, maxReputation: 499, dailyReportLimit: 8, color: '#A78BFA', glowColor: '#7C3AED', perks: ['view_profiles'] },
  { badgeId: '5', level: 5, name: 'Patrulheiro Local', nameEN: 'Local Patroller', icon: '🛡️', minReputation: 500, maxReputation: 749, dailyReportLimit: 8, color: '#F59E0B', glowColor: '#D97706', perks: ['comments', 'public_profile'] },
  { badgeId: '6', level: 6, name: 'Informante Confiável', nameEN: 'Trusted Informant', icon: '📡', minReputation: 750, maxReputation: 999, dailyReportLimit: 12, color: '#10B981', glowColor: '#047857', perks: ['wider_radius'] },
  { badgeId: '7', level: 7, name: 'Vigia Dedicado', nameEN: 'Dedicated Watchman', icon: '🔦', minReputation: 1000, maxReputation: 1499, dailyReportLimit: 12, color: '#3B82F6', glowColor: '#1D4ED8', perks: ['follow_users', 'feed'] },
  { badgeId: '8', level: 8, name: 'Sentinela de Bairro', nameEN: 'Neighborhood Sentinel', icon: '🏘️', minReputation: 1500, maxReputation: 1999, dailyReportLimit: 12, color: '#8B5CF6', glowColor: '#6D28D9', perks: ['12_reports'] },
  { badgeId: '9', level: 9, name: 'Guardião de Esquina', nameEN: 'Corner Guardian', icon: '🚦', minReputation: 2000, maxReputation: 2999, dailyReportLimit: 12, color: '#EC4899', glowColor: '#BE185D', perks: ['guardscan_3km'] },
  { badgeId: '10', level: 10, name: 'Sentinela Urbano', nameEN: 'Urban Sentinel', icon: '🌃', minReputation: 3000, maxReputation: 3999, dailyReportLimit: 12, color: '#14B8A6', glowColor: '#0D9488', perks: ['leaderboard'] },
  { badgeId: '11', level: 11, name: 'Protetor de Zona', nameEN: 'Zone Protector', icon: '🗺️', minReputation: 4000, maxReputation: 4999, dailyReportLimit: 18, color: '#F97316', glowColor: '#EA580C', perks: ['18_reports'] },
  { badgeId: '12', level: 12, name: 'Agente Comunitário', nameEN: 'Community Agent', icon: '🤝', minReputation: 5000, maxReputation: 6499, dailyReportLimit: 18, color: '#06B6D4', glowColor: '#0891B2', perks: ['community_alerts'] },
  { badgeId: '13', level: 13, name: 'Vigilante de Área', nameEN: 'Area Vigilante', icon: '🔭', minReputation: 6500, maxReputation: 7999, dailyReportLimit: 18, color: '#84CC16', glowColor: '#65A30D', perks: ['highlighted_markers'] },
  { badgeId: '14', level: 14, name: 'Guardião da Vizinhança', nameEN: 'Neighborhood Guardian', icon: '🏰', minReputation: 8000, maxReputation: 9999, dailyReportLimit: 18, color: '#EAB308', glowColor: '#CA8A04', perks: ['guardscan_4km'] },
  { badgeId: '15', level: 15, name: 'Escudo Popular', nameEN: "People's Shield", icon: '⚔️', minReputation: 10000, maxReputation: 12499, dailyReportLimit: 18, color: '#EF4444', glowColor: '#DC2626', perks: ['auto_boost_feed'] },
  { badgeId: '16', level: 16, name: 'Sentinela Avançado', nameEN: 'Advanced Sentinel', icon: '🎯', minReputation: 12500, maxReputation: 14999, dailyReportLimit: 25, color: '#8B5CF6', glowColor: '#7C3AED', perks: ['25_reports', 'area_analytics'] },
  { badgeId: '17', level: 17, name: 'Defensor Regional', nameEN: 'Regional Defender', icon: '🦅', minReputation: 15000, maxReputation: 17999, dailyReportLimit: 25, color: '#0EA5E9', glowColor: '#0284C7', perks: ['heatmap'] },
  { badgeId: '18', level: 18, name: 'Protetor Regional', nameEN: 'Regional Protector', icon: '🌐', minReputation: 18000, maxReputation: 21999, dailyReportLimit: 25, color: '#22D3EE', glowColor: '#06B6D4', perks: ['route_safety'] },
  { badgeId: '19', level: 19, name: 'Guardião de Elite', nameEN: 'Elite Guardian', icon: '💎', minReputation: 22000, maxReputation: 25999, dailyReportLimit: 25, color: '#A855F7', glowColor: '#9333EA', perks: ['priority_weight'] },
  { badgeId: '20', level: 20, name: 'Sentinela de Ouro', nameEN: 'Gold Sentinel', icon: '🏅', minReputation: 26000, maxReputation: 30999, dailyReportLimit: 25, color: '#FBBF24', glowColor: '#F59E0B', perks: ['guardscan_5km'] },
  { badgeId: '21', level: 21, name: 'Vigilante Supremo', nameEN: 'Supreme Vigilante', icon: '⭐', minReputation: 31000, maxReputation: 35999, dailyReportLimit: 35, color: '#FCD34D', glowColor: '#FBBF24', perks: ['35_reports', 'mentor_badge'] },
  { badgeId: '22', level: 22, name: 'Protetor Supremo', nameEN: 'Supreme Protector', icon: '🌟', minReputation: 36000, maxReputation: 41999, dailyReportLimit: 35, color: '#FB923C', glowColor: '#F97316', perks: ['mentor_users'] },
  { badgeId: '23', level: 23, name: 'Guardião Máximo', nameEN: 'Maximum Guardian', icon: '👑', minReputation: 42000, maxReputation: 49999, dailyReportLimit: 35, color: '#F472B6', glowColor: '#EC4899', perks: ['pinned_reports_8h'] },
  { badgeId: '24', level: 24, name: 'Comandante de Alerta', nameEN: 'Alert Commander', icon: '🎖️', minReputation: 50000, maxReputation: 57999, dailyReportLimit: 35, color: '#E879F9', glowColor: '#D946EF', perks: ['area_push_notifs'] },
  { badgeId: '25', level: 25, name: 'Sentinela Lendário', nameEN: 'Legendary Sentinel', icon: '🔱', minReputation: 58000, maxReputation: 65999, dailyReportLimit: 35, color: '#C084FC', glowColor: '#A855F7', perks: ['custom_marker'] },
  { badgeId: '26', level: 26, name: 'Defensor Lendário', nameEN: 'Legendary Defender', icon: '🛡️', minReputation: 66000, maxReputation: 74999, dailyReportLimit: 50, color: '#818CF8', glowColor: '#6366F1', perks: ['50_reports'] },
  { badgeId: '27', level: 27, name: 'Protetor Absoluto', nameEN: 'Absolute Protector', icon: '⚡', minReputation: 75000, maxReputation: 84999, dailyReportLimit: 50, color: '#38BDF8', glowColor: '#0EA5E9', perks: ['animated_badge'] },
  { badgeId: '28', level: 28, name: 'Guardião Imortal', nameEN: 'Immortal Guardian', icon: '🔥', minReputation: 85000, maxReputation: 92999, dailyReportLimit: 50, color: '#FB7185', glowColor: '#F43F5E', perks: ['custom_alert_sound'] },
  { badgeId: '29', level: 29, name: 'Mestre da Atenção', nameEN: 'Attention Expert', icon: '💫', minReputation: 93000, maxReputation: 99999, dailyReportLimit: 50, color: '#FACC15', glowColor: '#EAB308', perks: ['all_standard_perks'] },
  { badgeId: '30', level: 30, name: 'Attention Master', nameEN: 'Attention Master', icon: '🏆', minReputation: 100000, maxReputation: 199999, dailyReportLimit: 50, color: '#FFD700', glowColor: '#FFA500', perks: ['elite_marker', 'all_features', 'max_standard_level'] },
  { badgeId: 'guardian', level: 31, name: 'Guardian', nameEN: 'Guardian', icon: '🛡️', minReputation: 200000, maxReputation: null, dailyReportLimit: -1, color: '#00FFAA', glowColor: '#00FF88', perks: ['verify_incidents', 'review_reports', 'remove_incidents', 'moderation_dashboard', 'unlimited_reports', 'instant_verification', 'animated_guardian_badge', 'area_statistics'] },
];

export function getBadgeForLevel(level: number): Badge {
  return BADGES.find((b) => b.level === level) ?? BADGES[0];
}

export function getBadgeForReputation(reputation: number): Badge {
  if (reputation >= 200000) return BADGES[BADGES.length - 1];
  for (let i = BADGES.length - 2; i >= 0; i--) {
    if (reputation >= BADGES[i].minReputation) return BADGES[i];
  }
  return BADGES[0];
}

export function getProgressToNextLevel(reputation: number, currentBadge: Badge): number {
  if (currentBadge.maxReputation === null) return 1;
  const range = currentBadge.maxReputation - currentBadge.minReputation + 1;
  const progress = reputation - currentBadge.minReputation;
  return Math.min(progress / range, 1);
}

File: alert-flutter/lib/data/badges.dart
import 'dart:ui';

Color _hex(String hex) => Color(int.parse(hex.replaceFirst('#', 'FF'), radix: 16));

class Badge {
  final String badgeId;
  final int level;
  final String name;
  final String nameEN;
  final String icon;
  final int minReputation;
  final int? maxReputation;
  final int dailyReportLimit;
  final Color color;
  final Color glowColor;

  const Badge._({
    required this.badgeId,
    required this.level,
    required this.name,
    required this.nameEN,
    required this.icon,
    required this.minReputation,
    this.maxReputation,
    required this.dailyReportLimit,
    required this.color,
    required this.glowColor,
  });
}

final allBadges = [
  Badge._(badgeId: '0',  level: 0,  name: 'Observador Iniciante',     nameEN: 'Beginner observer',      icon: '👀',     minReputation: 0,      maxReputation: 49,     dailyReportLimit: 5,  color: _hex('#6B7280'), glowColor: _hex('#374151')),
  Badge._(badgeId: '1',  level: 1,  name: 'Vigia Desperto',           nameEN: 'Awakened watcher',        icon: '👁️‍🗨️', minReputation: 50,     maxReputation: 99,     dailyReportLimit: 5,  color: _hex('#9CA3AF'), glowColor: _hex('#4B5563')),
  Badge._(badgeId: '2',  level: 2,  name: 'Batedor Novato',           nameEN: 'Rookie scout',            icon: '🧭',     minReputation: 100,    maxReputation: 199,    dailyReportLimit: 5,  color: _hex('#60A5FA'), glowColor: _hex('#2563EB')),
  Badge._(badgeId: '3',  level: 3,  name: 'Repórter de Rua',          nameEN: 'Street reporter',         icon: '📰',     minReputation: 200,    maxReputation: 349,    dailyReportLimit: 8,  color: _hex('#34D399'), glowColor: _hex('#059669')),
  Badge._(badgeId: '4',  level: 4,  name: 'Olheiro do Bairro',        nameEN: 'Neighborhood lookout',    icon: '🏘️',    minReputation: 350,    maxReputation: 499,    dailyReportLimit: 8,  color: _hex('#A78BFA'), glowColor: _hex('#7C3AED')),
  Badge._(badgeId: '5',  level: 5,  name: 'Ronda Noturna',            nameEN: 'Night patrol',            icon: '🔦',     minReputation: 500,    maxReputation: 749,    dailyReportLimit: 8,  color: _hex('#F59E0B'), glowColor: _hex('#D97706')),
  Badge._(badgeId: '6',  level: 6,  name: 'Vigia de Sinais',          nameEN: 'Signal watcher',          icon: '📡',     minReputation: 750,    maxReputation: 999,    dailyReportLimit: 12, color: _hex('#10B981'), glowColor: _hex('#047857')),
  Badge._(badgeId: '7',  level: 7,  name: 'Guarda da Torre',          nameEN: 'Tower guard',             icon: '🗼',     minReputation: 1000,   maxReputation: 1499,   dailyReportLimit: 12, color: _hex('#3B82F6'), glowColor: _hex('#1D4ED8')),
  Badge._(badgeId: '8',  level: 8,  name: 'Sentinela de Ferro',       nameEN: 'Iron sentinel',           icon: '🔩',     minReputation: 1500,   maxReputation: 1999,   dailyReportLimit: 12, color: _hex('#8B5CF6'), glowColor: _hex('#6D28D9')),
  Badge._(badgeId: '9',  level: 9,  name: 'Intendente de Rua',        nameEN: 'Street warden',           icon: '🚧',     minReputation: 2000,   maxReputation: 2999,   dailyReportLimit: 12, color: _hex('#EC4899'), glowColor: _hex('#BE185D')),
  Badge._(badgeId: '10', level: 10, name: 'Vigia da Cidade',          nameEN: 'City watchman',           icon: '🌆',     minReputation: 3000,   maxReputation: 3999,   dailyReportLimit: 12, color: _hex('#14B8A6'), glowColor: _hex('#0D9488')),
  Badge._(badgeId: '11', level: 11, name: 'Defensor da Lei',          nameEN: 'Law enforcer',            icon: '👮',     minReputation: 4000,   maxReputation: 4999,   dailyReportLimit: 18, color: _hex('#F97316'), glowColor: _hex('#EA580C')),
  Badge._(badgeId: '12', level: 12, name: 'Agente de Campo',          nameEN: 'Field agent',             icon: '🕵️',    minReputation: 5000,   maxReputation: 6499,   dailyReportLimit: 18, color: _hex('#06B6D4'), glowColor: _hex('#0891B2')),
  Badge._(badgeId: '13', level: 13, name: 'Falcão Vigilante',         nameEN: 'Vigilant hawk',           icon: '🦅',     minReputation: 6500,   maxReputation: 7999,   dailyReportLimit: 18, color: _hex('#84CC16'), glowColor: _hex('#65A30D')),
  Badge._(badgeId: '14', level: 14, name: 'Protetor do Distrito',     nameEN: 'District protector',      icon: '🏛️',    minReputation: 8000,   maxReputation: 9999,   dailyReportLimit: 18, color: _hex('#EAB308'), glowColor: _hex('#CA8A04')),
  Badge._(badgeId: '15', level: 15, name: 'Porta-Escudo',             nameEN: 'Shield bearer',           icon: '🛡️',    minReputation: 10000,  maxReputation: 12499,  dailyReportLimit: 18, color: _hex('#EF4444'), glowColor: _hex('#DC2626')),
  Badge._(badgeId: '16', level: 16, name: 'Observador Tático',        nameEN: 'Tactical observer',       icon: '🎯',     minReputation: 12500,  maxReputation: 14999,  dailyReportLimit: 25, color: _hex('#8B5CF6'), glowColor: _hex('#7C3AED')),
  Badge._(badgeId: '17', level: 17, name: 'Comandante de Zona',       nameEN: 'Zone commander',          icon: '📯',     minReputation: 15000,  maxReputation: 17999,  dailyReportLimit: 25, color: _hex('#0EA5E9'), glowColor: _hex('#0284C7')),
  Badge._(badgeId: '18', level: 18, name: 'Oráculo da Segurança',     nameEN: 'Security oracle',         icon: '🔮',     minReputation: 18000,  maxReputation: 21999,  dailyReportLimit: 25, color: _hex('#22D3EE'), glowColor: _hex('#06B6D4')),
  Badge._(badgeId: '19', level: 19, name: 'Vigia de Elite',           nameEN: 'Elite watcher',           icon: '🔭',     minReputation: 22000,  maxReputation: 25999,  dailyReportLimit: 25, color: _hex('#A855F7'), glowColor: _hex('#9333EA')),
  Badge._(badgeId: '20', level: 20, name: 'Guardião de Ouro',         nameEN: 'Gold guardian',           icon: '🥇',     minReputation: 26000,  maxReputation: 30999,  dailyReportLimit: 25, color: _hex('#FBBF24'), glowColor: _hex('#F59E0B')),
  Badge._(badgeId: '21', level: 21, name: 'Sentinela Estelar',        nameEN: 'Star sentinel',           icon: '⭐',     minReputation: 31000,  maxReputation: 35999,  dailyReportLimit: 35, color: _hex('#FCD34D'), glowColor: _hex('#FBBF24')),
  Badge._(badgeId: '22', level: 22, name: 'Protetor Supremo',         nameEN: 'Supreme protector',       icon: '🦾',     minReputation: 36000,  maxReputation: 41999,  dailyReportLimit: 35, color: _hex('#FB923C'), glowColor: _hex('#F97316')),
  Badge._(badgeId: '23', level: 23, name: 'Guardião da Coroa',        nameEN: 'Crown warden',            icon: '👑',     minReputation: 42000,  maxReputation: 49999,  dailyReportLimit: 35, color: _hex('#F472B6'), glowColor: _hex('#EC4899')),
  Badge._(badgeId: '24', level: 24, name: 'Chefe da Vigília',         nameEN: 'Chief of watch',          icon: '📜',     minReputation: 50000,  maxReputation: 57999,  dailyReportLimit: 35, color: _hex('#E879F9'), glowColor: _hex('#D946EF')),
  Badge._(badgeId: '25', level: 25, name: 'Guardião Lendário',        nameEN: 'Legendary guardian',      icon: '🌟',     minReputation: 58000,  maxReputation: 65999,  dailyReportLimit: 35, color: _hex('#C084FC'), glowColor: _hex('#A855F7')),
  Badge._(badgeId: '26', level: 26, name: 'Escudo Mítico',            nameEN: 'Mythic shield',           icon: '🧱',     minReputation: 66000,  maxReputation: 74999,  dailyReportLimit: 50, color: _hex('#818CF8'), glowColor: _hex('#6366F1')),
  Badge._(badgeId: '27', level: 27, name: 'Guardião do Trovão',       nameEN: 'Thunder guardian',        icon: '⚡',     minReputation: 75000,  maxReputation: 84999,  dailyReportLimit: 50, color: _hex('#38BDF8'), glowColor: _hex('#0EA5E9')),
  Badge._(badgeId: '28', level: 28, name: 'Sentinela Eterna',         nameEN: 'Eternal sentinel',        icon: '🕰️',    minReputation: 85000,  maxReputation: 92999,  dailyReportLimit: 50, color: _hex('#FB7185'), glowColor: _hex('#F43F5E')),
  Badge._(badgeId: '29', level: 29, name: 'Vigia Ômega',              nameEN: 'Omega watcher',           icon: '⭕',     minReputation: 93000,  maxReputation: 99999,  dailyReportLimit: 50, color: _hex('#FACC15'), glowColor: _hex('#EAB308')),
  Badge._(badgeId: '30', level: 30, name: 'Guardião Grão-Mestre',     nameEN: 'Grand master guardian',   icon: '🏆',     minReputation: 100000, maxReputation: 199999, dailyReportLimit: 50, color: _hex('#FFD700'), glowColor: _hex('#FFA500')),
  Badge._(badgeId: 'guardian', level: 31, name: 'Guardião Supremo',   nameEN: 'Supreme guardian',        icon: '🔐',     minReputation: 200000, maxReputation: null,   dailyReportLimit: -1, color: _hex('#00FFAA'), glowColor: _hex('#00FF88')),
];

Badge getBadgeForLevel(int level) {
  return allBadges.firstWhere((b) => b.level == level, orElse: () => allBadges.first);
}

Badge getBadgeForReputation(int reputation) {
  if (reputation >= 200000) return allBadges.last;
  for (int i = allBadges.length - 2; i >= 0; i--) {
    if (reputation >= allBadges[i].minReputation) return allBadges[i];
  }
  return allBadges.first;
}

double getProgressToNextLevel(int reputation, Badge currentBadge) {
  if (currentBadge.maxReputation == null) return 1.0;
  final range = currentBadge.maxReputation! - currentBadge.minReputation + 1;
  final progress = reputation - currentBadge.minReputation;
  return (progress / range).clamp(0.0, 1.0);
}
